### PR TITLE
Jds mempool enhancement 3

### PR DESCRIPTION
**Pull request by lorbax**
apply part of the suggestions to this PR
[https://github.com/stratum-mining/stratum/pull/772](https://github.com/stratum-mining/stratum/pull/772)

### CODE REVIEW

File: roles/jd-server/src/lib/mempool/mod.rs [hunk 1]
```diff
@@ -58,14 +59,34 @@ impl JDsMempool {
...
+        let transactions = add_txs_to_mempool_inner.unknown_transactions;
+        let client = self_
+            .safe_lock(|a| a.get_client())
+            .map_err(|e| JdsMempoolError::PoisonLock(e.to_string()))?
+            .ok_or(JdsMempoolError::NoClient)?;
+        // fill in the mempool the transactions id in the mempool with the full transactions
+        // retrieved from the jd client
+        for txid in txids {
+            let transaction = client
+                .get_raw_transaction(&txid.to_string(), None)
```
Comment (Fi3) [inline roles/jd-server/src/lib/mempool/mod.rs]: before fetching txs you should check if they are already in the mempool

Comment (Fi3) [reply roles/jd-server/src/lib/mempool/mod.rs]: another option could  be to use an HashSet right after the receiver

Comment (lorbax) [reply roles/jd-server/src/lib/mempool/mod.rs]: I am struggling to figure out a useful case for HashSet. If you propose to use it instead HashMap, then you cant store transactions only by their IDs. What do you mean with "right after the receiver"? Can you be a bit more specific? ^^

BTW already fixed in [this commit](https://github.com/lorbax/stratum/commit/d99d438268660bf31bc4d10cd7c30f7244887141)

File: roles/jd-server/src/lib/job_declarator/mod.rs [hunk 2]
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
...
+            .unwrap();
+        let sender_add_txs_to_mempool = add_txs_to_mempool.sender_add_txs_to_mempool;
+        let add_txs_to_mempool_inner = add_txs_to_mempool.add_txs_to_mempool_inner;
+        let _ = sender_add_txs_to_mempool
+            .send(add_txs_to_mempool_inner)
+            .await;
+        // the trasnactions sent to the mempool can be freed
+        let _ = self_mutex.safe_lock(|a| {
+            a.add_txs_to_mempool.add_txs_to_mempool_inner = AddTrasactionsToMempoolInner {
+                known_transactions: vec![],
```
Comment (Fi3) [inline roles/jd-server/src/lib/job_declarator/mod.rs:173]: better to use clear, no need to deallocate if we are going to fill it again

File: roles/jd-server/src/lib/job_declarator/mod.rs [hunk 3]
```diff
@@ -231,9 +222,54 @@ impl JobDeclaratorDownstream {
...
+                                    JobDeclaration::ProvideMissingTransactions(_) => {
+                                        info!("Send message: PMT. Updating the JDS mempool.");
+                                        Self::send_txs_to_mempool(self_mutex.clone()).await;
+                                    }
+                                    JobDeclaration::ProvideMissingTransactionsSuccess(_) => {
+                                        error!("Send unexpected PMTS");
+                                    }
+                                    JobDeclaration::SubmitSolution(_) => todo!(),
+                                }
+                                Self::send(self_mutex.clone(), m).await.unwrap();
```
Comment (Fi3) [inline roles/jd-server/src/lib/job_declarator/mod.rs:272]: this always send the message also for the error cases

Comment (Fi3) [reply roles/jd-server/src/lib/job_declarator/mod.rs:272]: if we get one of the error cases above something really strange is happening better to close the connection. Downstream will decide what to do

File: roles/jd-server/src/lib/job_declarator/mod.rs [hunk 4]
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
...
             };
         }
-        let block: Block =
-            roles_logic_sv2::utils::BlockCreator::new(last_declare, transactions_list, message)
-                .into();
-        Ok(hex::encode(serialize(&block)))
+        Ok(transactions_list)
+    }
+
+    async fn send_txs_to_mempool(self_mutex: Arc<Mutex<Self>>) {
```
Comment (Fi3) [inline roles/jd-server/src/lib/job_declarator/mod.rs:161]: I think that you can use an &Arc<Mutex<Seld>> here

File: roles/jd-server/src/lib/job_declarator/mod.rs [hunk 5]
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
...
+            .map_err(|e| Box::new(JdsError::PoisonLock(e.to_string())))?;
+        let last_declare = last_declare_.ok_or(Box::new(JdsError::NoLastDeclaredJob))?;
+        let transactions_list = Self::are_all_job_transactions_present(self_mutex)?;
+        let block: Block =
+            roles_logic_sv2::utils::BlockCreator::new(last_declare, transactions_list, message)
+                .into();
+        Ok(hex::encode(serialize(&block)))
+    }
+
+    fn are_all_job_transactions_present(
```
Comment (Fi3) [inline roles/jd-server/src/lib/job_declarator/mod.rs:126]: change the function name i something else, this seems just a condition check but it do collect txs.

you can use colletc_txs_for_job

File: roles/jd-server/src/lib/job_declarator/mod.rs [hunk 6]
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
...
+        Ok(transactions_list)
+    }
+
+    async fn send_txs_to_mempool(self_mutex: Arc<Mutex<Self>>) {
+        let add_txs_to_mempool = self_mutex
+            .safe_lock(|a| a.add_txs_to_mempool.clone())
+            .unwrap();
+        let sender_add_txs_to_mempool = add_txs_to_mempool.sender_add_txs_to_mempool;
+        let add_txs_to_mempool_inner = add_txs_to_mempool.add_txs_to_mempool_inner;
+        let _ = sender_add_txs_to_mempool
```
Comment (Fi3) [inline roles/jd-server/src/lib/job_declarator/mod.rs:167]: why you ignore the error here?

File: roles/jd-server/src/lib/job_declarator/mod.rs [hunk 7]
```diff
@@ -90,113 +100,94 @@ impl JobDeclaratorDownstream {
...
+            .send(add_txs_to_mempool_inner)
+            .await;
+        // the trasnactions sent to the mempool can be freed
+        let _ = self_mutex.safe_lock(|a| {
```
Comment (Fi3) [inline roles/jd-server/src/lib/job_declarator/mod.rs:171]: same

File: roles/jd-server/src/lib/job_declarator/mod.rs [hunk 8]
```diff
@@ -247,55 +283,102 @@ impl JobDeclaratorDownstream {
...
+                                                let retrieve_transactions =
+                                                    AddTrasactionsToMempoolInner {
+                                                        known_transactions,
+                                                        unknown_transactions: Vec::new(),
+                                                    };
+                                                let mempool = self_mutex
+                                                    .clone()
+                                                    .safe_lock(|a| a.mempool.clone())
+                                                    .unwrap();
+                                                tokio::select! {
```
Comment (Fi3) [inline roles/jd-server/src/lib/job_declarator/mod.rs:328]: seems that you want to use a tokio::timeout here

File: roles/jd-server/src/lib/job_declarator/mod.rs [hunk 9]
```diff
@@ -408,6 +491,8 @@ impl JobDeclarator {
                     sender.clone(),
                     &config,
                     mempool.clone(),
+                    // each downstream has its own sender (multi producer single consumer)
```
Comment (Fi3) [inline roles/jd-server/src/lib/job_declarator/mod.rs:494]: remove this comment you see it from the type

**Pull request merged**